File: BOJ/#1541 잃어버린 괄호/잃어버린 괄호.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;

int main() {
    string s; cin>>s;

    vector<int> nums;
    vector<char> ops;

    int idx = 0;
    for(int i=0; i<s.length(); i++) {
        if(!isdigit(s[i])) {
            nums.push_back(stoi(s.substr(idx, i-idx)));
            idx = i+1;

            ops.push_back(s[i]);
        }
    }
    nums.push_back(stoi(s.substr(idx)));

    int answer = nums[0];
    for(int i=0; i<ops.size(); i++) {
        // +면 그냥 더하기
        if(ops[i] == '+') answer += nums[i+1];
        // -면 다음 -까지 다 더한 후 빼기
        else {
            int tmp = nums[i+1];
            for(int j=i+1; j<ops.size(); j++) {
                if(ops[j] == '+') tmp += nums[j+1];
                else {
                    answer -= tmp;
                    tmp = -1;
                    i = j-1;
                    break;
                }
            }

            if(tmp != -1) {
                answer -= tmp;
                break;
            }
        }
    }

    cout<<answer;
}

File: BOJ/#1541 잃어버린 괄호/잃어버린 괄호.py
s = input()
nums,ops = [],[]

idx = 0
for i in range(len(s)):
    if s[i]=='+' or s[i]=='-':
        nums.append(int(s[idx:i]))
        idx = i+1

        ops.append(s[i])
nums.append(int(s[idx:]))

answer = nums[0]
i = 0
while i < len(ops):
    if ops[i] == '+': answer += nums[i+1]
    else:
        tmp = nums[i+1]
        for j in range(i+1, len(ops)):
            if ops[j] == '+': tmp += nums[j+1]
            else:
                answer -= tmp
                tmp = -1
                i = j-1
                break

        if tmp != -1:
            answer -= tmp
            break

    i += 1

print(answer)
